torch/fx/experimental/symbolic_shapes.py

```diff
@@ -978,8 +1031,10 @@ def evaluate_expr(self, expr: "sympy.Expr"):
         """
         Given an expression, evaluates it, adding guards if necessary
         """
-        if len(expr.free_symbols) == 0:
-            return expr
+        # NB: do not short-circuit here without evaluating as we might need to
+        # compute FloorDiv exprs in some cases. That is, anything not simplified
+        # by FloorDiv.eval.
+        # See NOTE [ SymPy eval and assumptions ]
         expr = self.simplify(expr)
```

Comment:
Do the len(expr.free_symbols) test here then? But I still think there's something fishy going on here. How come everything else in sympy was actually fully evaluated at this point without simplify...

Reply:
- `simplify` won't work because we don't define some cases for our `FloorDiv`, it won't know what to do:
```py
[ins] In [251]: class Moo(sympy.Function):
           ...:     @classmethod
           ...:     def eval(self, x, y):
           ...:         if x < 2:
           ...:             return x + y
           ...:

[ins] In [252]: Moo(2,3)
Out[252]: Moo(2, 3)

[ins] In [253]: Moo(1,3)
Out[253]: 4

[ins] In [254]: sympy.simplify(Moo(2,3))
Out[254]: Moo(2, 3)
```

- that's by design: we're subclassing `Function` which is for symbolic exprs (AKA exprs you want to keep uneval'd in some cases)

- some things in sympy subclass `Function` too, but many subclass `Expr` instead, like `Add`. Those that are symbolic will fail to simplify too:

```py
[ins] In [263]: sympy.simplify(sympy.log(sympy.sympify(8)/3, 2))
Out[263]: -log(3)/log(2) + 3
```

- depending on how you define eval, the advantage of symbolic is that you can simplify without compute or just with Symbols, but you still need to eval somewhere. we eval in `_maybe_evaluate_static` by pattern-matching on `FloorDiv`

Reply:
bumping this comment

Reply:
Aha! So `simplify` also automatically calls `doit`, which you can define on a class for full eval. There we can eval and return after `simplify` as before if there are no free vars. I'll try implementing that.

Reply:
Added doit to have more control over eval + some tests. PTAL

Reply:
Also: The `log` snippet posted above in this thread has no free vars either. And it's not fully eval'd. So the eval strategy is up to however you want it to work.

Reply:
That's because Moo is not total. All of the functions we are defining are total

Reply:
Right, but Moo is essentially FloorDiv. Note that we sometimes return None from FloorDiv as well. Which is why we need to fully eval it somewhere else.

Feel free to ping me offline if you want me to explain further.

Reply:
FloorDiv should be total when all the inputs are Integer (aka known literals)

Reply:
It does simplify in the Integer case as is:

```
[nav] In [15]: FloorDiv(6, (FloorDiv(12, 6)))
Out[15]: 3
```

The reason _additional_ eval is needed is that `FloorDiv` might return None in some cases (when the `eval` classmethod returns None). So we need to somehow convert `FloorDiv` to an actual value in `evaluate_expr`. Using `doit` is one way of doing that.

I also mentioned it before, but the reason we need to return a concrete value from `evaluate_expr` is that it's used in `guard_float`/`guard_int`. E.g., `guard_int` will attempt to call `int(FloorDiv(...))`, which won't work.

Reply:
It should simplify even with mixed arguments

Reply:
I suppose one could hypothetically argue there would be precision could be lost for some ops so they shouldn't evaluate, but I don't feel this problem applies to floordiv

Reply:
> It should simplify even with mixed arguments

But I don't see how this could work. Either:
- we need to make FloorDiv total for all inputs (it was mentioned before that we don't want to do that)
- OR we could move the code I have in doit into FloorDiv and apply it for LHS and RHS sub-exprs (eval if no free vars).

Are you okay with either of these solutions? Or do you see a different way to approach this?

Reply:
It seems clear to me that FloorDiv's eval function should actually evaluate when the inputs are Integers or Floats. It seems crazy to not evaluate with a Float inputs.

Why do you think we don't want to make FloorDiv total for this case? Are you talking about your comment https://github.com/pytorch/pytorch/pull/90906#discussion_r1064048676 ? I reread the Sympy documentation on the matter and it doesn't seem to say anything about the Integer/Floating input case.

Reply:
- I can add this code to `FloorDiv.eval`, then it'll eval Float/Integer inputs. Problem solved. _I'll push this fix shortly._
```py
if isinstance(base, (sympy.Integer, sympy.Float)) and isinstance(divisor, (sympy.Integer, sympy.Float)):
    return sympy.floor(base / divisor)
```

- I was concerned about funcs that _have no free vars_ and _do not simplify fully_. E.g., `log`:
`sympy.log(sympy.sympify(8)/3, 2)`. Turns out, however, it doesn't simplify even with `.doit(deep=True)`, so no point in keeping our `doit` code around, I guess. I might be wrong, but I feel like these should be pretty rare, because most arithmetic funcs are not defined by wrapping Function in sympy. I think we can address this later if there are problems.

Reply:
Updated. LMK if there's anything else that needs changing. Otherwise, please stamp it so I could merge. (I'll re-read the whole diff once again before merging to make sure we didn't miss any minor issues.)

Reply:
At the end of the day, the degree of automatic evaluation that you have in `eval` is up to you, and it depends on what you want out of your function. If evaluating too little is an annoyance that makes you have to call `doit` too much, then that may be a sign that you should move the evaluation to `eval` to happen automatically. 

I've tried to outline some of the pitfalls to avoid here https://docs.sympy.org/latest/guides/custom-functions.html#defining-automatic-evaluation-with-eval, but even if you don't follow those guidelines, SymPy will still support whatever degree of evaluation you choose to implement. That doc tends to caution towards less automatic evaluation because you can always go in one direction, to evaluate something that doesn't evaluate automatically, but doing the reverse---preventing something from evaluating automatically---is harder (it can only be done with `evaluate=False`, which is fragile). This all comes from the experience of the sorts of issues we've run into in SymPy itself. I do recommend evaluating on explicit numerical inputs (when the value is itself a number) because it's rare to not want that. 

FWIW, I personally think `log` does too much automatic evaluation on rational numbers (unfortunately, quite a few functions in SymPy itself don't exactly comply with the advice given in that document). I would much rather it only evaluated when the result doesn't have any logs, and leave the rest to simplification functions like `expand_log` and `logcombine`.